### ISSUE TITLE
[Feature Request] TaskHandler should return the task_id if persistent.

### ISSUE THREAD

**Issue by kovitikus** [evennia/evennia#2370]
#### Is your feature request related to a problem? Please describe.
There is currently no way to manage a persistent `utils.delay()` after a server has been reloaded.

#### Describe the solution you'd like
The TaskHandler `add()` method should return the `task_id` if the `persistent` kwarg is set to `True`.

#### Additional context
There is currently a `remove()` method that takes the `task_id` as an argument, but there is no way for the developer to gain access to the `task_id` when creating a `utils.delay()`. This change would greatly improve the usability of `utils.delay()` across a reload.

Chat conversation regarding this proposed change:
https://irclogs.jackgrigg.com/irc.freenode.net/evennia/2021-04-07#i_4710369

**Comment by davewiththenicehat** [comment]
For reference:
Branch off of develop, but merge back into master? Ref: "Griatch | Any api changes to utils.delay would happen in develop though". Won't that merge all the changes in develop into master?
TaskHandler.add, if persistent=True, deferLater, than return the taskid. We are not concerned with access the the deferral instance because moving forward we have to assume it does not exist due to a restart.

If that is all that is needed, this is a simple resolution. As far as I can tell that is it. All the other required features already exist, they just lack the key reference to the global task handler list.

There is a minor cost to this, in that if someone is creating a persistent task and calling the deferred instance it will throw an exception because int is not a deferred. It MAY break some people's code.

Otherwise, I'd like to implement this if I am understanding correctly. This is a very simple fix that allows me to use the task handler properly.

**Comment by kovitikus** [comment]
Yeah, I was a bit confused on the develop and master branch comment he made regarding these changes.

Yes, I think this will break some folks code who are currently setting persistent=True, but storing it in an ndb and attempting call cancel on it WITHOUT a server reload. Because their current code is expecting to have the deferred.

My suggestion here is to add a new kwarg that would instead force a return of the task_id instead of the deferred and leave the default behavior as is for both persistent and non-persistent.

**Comment by Griatch** [comment]
> Yes, I think this will break some folks code who are currently setting persistent=True, but storing it in an ndb and attempting call cancel on it WITHOUT a server reload. Because their current code is expecting to have the deferred.

The reason for returning deferreds is that one can chain them dynamically with additional callbacks/errbacks, but it depends on operating in-memory. Running a persistent delay with the expectation of also being able to operate directly on its deferred across reloads is incorrect code.

If one wanted to support something like this under one consistent API, maybe one could consider some sort of mock container object that wraps the `deferred` in a way so as to make it serializable (and thus saveable to the database) but have it recreate the actual deferred from the TaskHandler under the hood when you retrieve it. It would require quite a bit of Python magic to get something like that to work transparently, but I guess it could be possible.

**Comment by davewiththenicehat** [comment]
I've got to be misunderstanding if you think this would require work. I figured I would put it in code.
Don't I just want to call deferLater and return the task id?

**Comment by kovitikus** [comment]
> Running a persistent delay with the expectation of also being able to operate directly on its deferred across reloads is incorrect code.

What I actually meant here is that some folk may set their delay to persistent and attempt manipulations (chaining them) on it before the server is reloaded, not after. So these people would still need a deferred returned even when set to persistent.

My suggestion of adding an additional kwarg of `task_id` *(might have to be named differently)* to both `utils.delay()` and the task handler `add()` would allow them to return both the deferred and the task_id. Allowing the developer to still make use of the deferred in your chaining example, while also saving the task_id for later recall, even if it isn't persistent.

Of course the problem lies with the cleanup of the task_id. If it isn't persistent and the server gets reloaded, that task_id will still be stored in some attribute somewhere and the code may attempt to call upon it. In most normal cases the developer will clean up the task_id saved to the attribute within the callback or when calling `remove()`, since the task handler has no knowledge of the attribute this id is stored on.

As far as the advanced container stuff, I have no clue what any of that means, but it sounds like it's the most solid path forward if you feel it manages things better.

**Comment by davewiththenicehat** [comment]
It should be possible to make the callback argument iterable. Making it possible to add additional call backs to the TaskHandler entry. Since chaining is just another example of adding tasks it could be implemented also. Than you would make an additional handler for delay so you have a custom object to return that can survive a restart.
However I don't think it is worth the 40 or so hours it would take to make and test that class. I would not anticipate anyone using that feature.

Myself I see this as Griatch does. Persistent should not return an instance of a deferred. We can do everything needed for a delay with the id. You just get an instance of the global task handler, give it's methods the task id. In my mind that is very easy.
There are a number of things that can go wrong by referencing memory when the developer is expecting persistence.

If this project were at or past the 1.0 version mark my opinion would be drastically different. It is still below that mark. Changes such as returning an int rather than a deferred are acceptable for a beta project.

**Comment by davewiththenicehat** [comment]
It's still possible to chain a task onto a persistent delay.
in the add method leave a space for an additional list element.
`self.tasks[task_id] = (now + delta, callback, chain=list(), safe_args, safe_kwargs)`

Make a chain method
```python
TASK_Handler.chain(task_id, callback, *args, **kwargs)
    # create the task as a persistent task but do not call deferrLater
    self.tasks[task_id][2].append(chain_task_id)
```

do_task iterates through the chain id's calling their callback than deleting their entry in the global TaskHandler.

**Comment by kovitikus** [comment]
I'd much rather work with the `task_id` than with a deferred anyhow, so if you and Griatch think this solution is sound, you have my vote.

**Comment by kovitikus** [comment]
@davewiththenicehat  @Griatch 

If it's possible to do all operations, including chaining the defer with a method, will you be changing the TaskHandler to only return the `task_id` no matter if `persistent=True` or not?

Even in non-persistent tasks, I'd rather manage the task with the TaskHandler methods, so I'm not sure I see a need to return the deferred at all.

**Comment by davewiththenicehat** [comment]
I am confident I could write it in.
Though deferred has a lot of method. So task handler would have a get method to return a deferred.

It would also break anyone's current code that relies on the returned deferred instance.

**Comment by kovitikus** [comment]
Thanks. I think that dealing with deferred through the TaskHandler, rather than directly, feels more logical to me. That's sorta the point of the TaskHandler, to give a user-friendly interface for predetermined actions upon a deferred task.

Also, I think we're way past the point of worrying about breaking people's current code. They'll just have to be notified and hopefully fix it after the improvements are put out.

I'm really glad to see this getting improvements, because it does feel like the TaskHandler was half-finished.

**Comment by davewiththenicehat** [comment]
As I'm researching this it is not entirely the fault of who had written task handler.
For example twisted's deferLater method of getting the twist's cancel method from the clock did not work.
I had to rewrite deferLater to resolve it.

Before rewriting to return only task ids I am going to wait for Griatch to respond.
I have plenty more work to do while we wait.

**Comment by Griatch** [comment]
> For example twisted's deferLater method of getting the twist's cancel method from the clock did not work.
I had to rewrite deferLater to resolve it.

Not sure what you mean here - `deferred.cancel()` works just fine as far as I know? We use these tools a lot in Evennia and I've never had a need to rewrite something as fundamental as `deferLater` to work with it ...

I think that breaking backwards compatibility and return the id rather than the defererred is ok in this case since currently the taskhandler.remove is just not usable. 
Not sure `.chain` is really needed (at least if I understand it as some sort of custom thing) - as long as you get the deferred you can use that to chain as much as you want. 
The simplest way forward here may be to just allow the user a new kwarg to choose to return the deferred if desired from `.add`, and maybe some new method to be able to retrieve the id and/or deferred from the handler on-demand if they want to do something with them.

**Comment by davewiththenicehat** [comment]
I will move forward with returning only a task id.

I don't know that cancel is called on any deferrals within evennia. I looked and found none.

If I test with code:
```python:
deferal_inst = utils.delay(1, dummy_func, self.char1.dbref)
        deferal_inst.callback(True)
        deferal_inst.cancel()
        self.assertEqual(self.char1.ndb.dummy_var, 'dummy_func ran')
        self.char1.ndb.dummy_var = False
        _TASK_HANDLER.clock.advance(1)  # make time pass
```
I will run into an AlreadyCalledError. Because the callback has been called. Cancel though does not stop the scheduled task from firing.

Example of deferLater without comments:
```python
def deferLater(clock, delay, callable=None, *args, **kw):
    def deferLaterCancel(deferred):
        delayedCall.cancel()
    d = defer.Deferred(deferLaterCancel)
    if callable is not None:
        d.addCallback(lambda ignored: callable(*args, **kw))
    delayedCall = clock.callLater(delay, d.callback, None)
    return d
```

I am guessing that delayedCall no longer exists when the instance of deferLaterCancel is called.

I do know that if I get a reference of the delayedCall and overwrite the deferral instance's cancel method the call is properly canceled.
IE `d.cancel = delayedCall.cancel`

I am not done testing yet.

**Comment by kovitikus** [comment]
I found `.cancel()` in the following locations. Hope these results help out.

Scripts.
https://github.com/evennia/evennia/blob/master/evennia/scripts/scripts.py#L113
***
Portal testing.
https://github.com/evennia/evennia/blob/master/evennia/server/portal/tests.py#L279
***
Slow Exit contrib that will need to be updated, cause it will break with this change.
https://github.com/evennia/evennia/blob/master/evennia/contrib/slow_exit.py#L139

**Comment by Griatch** [comment]
@davewiththenicehat The cancel works fine. You are probably not testing in an environment where the reactor is running. 

This is what I tested: 

First place a simple method on the character typeclass:

```python
def send_message(self);
    self.msg("A delayed message!")
```

Then reload and in game:

    py defer = evennia.utils.delay(10, self.send_message)

After 10 seconds the message will pop up. Next:

    py defer = evennia.utils.delay(10, self.send_message);defer.cancel()

Now the message will not show up, since the deferred was canceled. Which is expected. If I tried to call it later than 10 seconds later, I'd get an already-called error, which is also correct. One would need to catch this or check with `defer.called` to be sure.

**Comment by davewiththenicehat** [comment]
To see the errors you would also need to call it early.
Seeing this I know cancel is working.
I'm reading defer.py to get an understanding of what is occurring.

In short calling a defer early is not simple. As calling it does not unscheduled it (because it is scheduled separately).
Than it errors with an already called if it has not been unscheduled. To boot deferLater does not expose the scheduler.

I have what I need. I'll figure it out.

**Comment by Griatch** [comment]
Closed in #2374
